File: lib/klog.c
/*
 * Unified kernel logging with early-boot support.
 * Falls back to raw COM1 writes until the serial driver is attached.
 */

#include "klog.h"
#include "../boot/constants.h"
#include "../drivers/serial.h"
#include "../lib/io.h"
#include "../lib/numfmt.h"

static enum klog_level current_level = KLOG_INFO;
static int serial_ready = 0;

static const char hex_digits[] = "0123456789ABCDEF";

static void klog_early_putc(char c) {
    io_outb(COM1_BASE, c);
}

static void klog_emit(const char *text) {
    if (!text) {
        return;
    }

    if (serial_ready) {
        kprint(text);
        return;
    }

    const char *p = text;
    while (*p) {
        klog_early_putc(*p++);
    }
}

static void klog_emit_line(const char *text) {
    if (text) {
        klog_emit(text);
    }
    if (serial_ready) {
        kprint("\n");
    } else {
        klog_early_putc('\n');
    }
}

void klog_init(void) {
    current_level = KLOG_INFO;
    serial_ready = 0;
}

void klog_attach_serial(void) {
    serial_ready = 1;
}

void klog_set_level(enum klog_level level) {
    current_level = level;
}

enum klog_level klog_get_level(void) {
    return current_level;
}

int klog_is_enabled(enum klog_level level) {
    return level <= current_level;
}

void klog_newline(void) {
    klog_emit_line(NULL);
}

void klog(enum klog_level level, const char *msg) {
    if (!klog_is_enabled(level)) {
        return;
    }
    klog_emit_line(msg);
}

void klog_raw(enum klog_level level, const char *msg) {
    if (!klog_is_enabled(level) || !msg) {
        return;
    }
    klog_emit(msg);
}

void klog_hex(enum klog_level level, uint64_t value) {
    if (!klog_is_enabled(level)) {
        return;
    }

    char buffer[19];
    buffer[0] = '0';
    buffer[1] = 'x';
    for (int i = 0; i < 16; i++) {
        buffer[2 + i] = hex_digits[(value >> ((15 - i) * 4)) & 0xF];
    }
    buffer[18] = '\0';
    klog_raw(level, buffer);
}

void klog_decimal(enum klog_level level, uint64_t value) {
    if (!klog_is_enabled(level)) {
        return;
    }

    char buffer[32];
    if (numfmt_u64_to_decimal(value, buffer, sizeof(buffer)) == 0) {
        buffer[0] = '0';
        buffer[1] = '\0';
    }
    klog_raw(level, buffer);
}

File: lib/src/klog.rs
use core::ffi::c_int;
use core::fmt;
use core::sync::atomic::{AtomicU8, Ordering};

use crate::init_flag::InitFlag;
use crate::ports::COM1;

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum KlogLevel {
    Error = 0,
    Warn = 1,
    Info = 2,
    Debug = 3,
    Trace = 4,
}

impl KlogLevel {
    fn from_raw(raw: u8) -> Self {
        match raw {
            0 => KlogLevel::Error,
            1 => KlogLevel::Warn,
            2 => KlogLevel::Info,
            3 => KlogLevel::Debug,
            _ => KlogLevel::Trace,
        }
    }
}

static CURRENT_LEVEL: AtomicU8 = AtomicU8::new(KlogLevel::Info as u8);
static SERIAL_READY: InitFlag = InitFlag::new();

#[inline(always)]
fn is_enabled(level: KlogLevel) -> bool {
    level as u8 <= CURRENT_LEVEL.load(Ordering::Relaxed)
}

#[inline(always)]
fn putc(byte: u8) {
    let _ready = SERIAL_READY.is_set_relaxed();
    unsafe { COM1.write(byte) }
}

fn write_bytes(bytes: &[u8]) {
    for &b in bytes {
        putc(b);
    }
}

pub(crate) fn log_line(level: KlogLevel, text: &str) {
    if !is_enabled(level) {
        return;
    }
    write_bytes(text.as_bytes());
    putc(b'\n');
}

pub fn is_enabled_level(level: KlogLevel) -> bool {
    is_enabled(level)
}

pub fn log_args(level: KlogLevel, args: fmt::Arguments<'_>) {
    if !is_enabled(level) {
        return;
    }
    struct KlogWriter;
    impl fmt::Write for KlogWriter {
        fn write_str(&mut self, s: &str) -> fmt::Result {
            write_bytes(s.as_bytes());
            Ok(())
        }
    }
    let _ = fmt::write(&mut KlogWriter, args);
    putc(b'\n');
}
pub fn klog_init() {
    CURRENT_LEVEL.store(KlogLevel::Info as u8, Ordering::Relaxed);
    SERIAL_READY.reset();
}
pub fn klog_attach_serial() {
    SERIAL_READY.mark_set();
}
pub fn klog_set_level(level: KlogLevel) {
    CURRENT_LEVEL.store(level as u8, Ordering::Relaxed);
}
pub fn klog_get_level() -> KlogLevel {
    KlogLevel::from_raw(CURRENT_LEVEL.load(Ordering::Relaxed))
}
pub fn klog_is_enabled(level: KlogLevel) -> c_int {
    if is_enabled(level) {
        1
    } else {
        0
    }
}
pub fn klog_newline() {
    putc(b'\n');
}

#[macro_export]
macro_rules! klog {
    ($level:expr, $($arg:tt)*) => {{
        $crate::klog::log_args($level, ::core::format_args!($($arg)*));
    }};
}

#[macro_export]
macro_rules! klog_error {
    ($($arg:tt)*) => {
        $crate::klog::log_args($crate::klog::KlogLevel::Error, ::core::format_args!($($arg)*))
    };
}

#[macro_export]
macro_rules! klog_warn {
    ($($arg:tt)*) => {
        $crate::klog::log_args($crate::klog::KlogLevel::Warn, ::core::format_args!($($arg)*))
    };
}

#[macro_export]
macro_rules! klog_info {
    ($($arg:tt)*) => {
        $crate::klog::log_args($crate::klog::KlogLevel::Info, ::core::format_args!($($arg)*))
    };
}

#[macro_export]
macro_rules! klog_debug {
    ($($arg:tt)*) => {
        $crate::klog::log_args($crate::klog::KlogLevel::Debug, ::core::format_args!($($arg)*))
    };
}

#[macro_export]
macro_rules! klog_trace {
    ($($arg:tt)*) => {
        $crate::klog::log_args($crate::klog::KlogLevel::Trace, ::core::format_args!($($arg)*))
    };
}
